aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -744,6 +746,65 @@ SparseTensor _coalesce_sparse_cpu(const SparseTensor& self) {
 }
 
 DEFINE_DISPATCH(sparse_mask_intersection_out_stub);
+DEFINE_DISPATCH(sparse_mask_projection_out_stub);
+
+using OptTensor = c10::optional<Tensor>;
+
+std::tuple<Tensor, Tensor, OptTensor> sparse_mask_like_prepare_sparse_inputs(
+    const std::string& method_name,
+    const Tensor& t,
+    const Tensor& mask) {
+  TORCH_CHECK(
+      mask.sizes().equals(t.sizes()),
+      method_name, "(): operands have incompatible sizes; self has size ",
+      t.sizes(),
+      " but mask has size ",
+      mask.sizes());
+
+  TORCH_CHECK(t.sparse_dim() == mask.sparse_dim(),
+              method_name, "(): the number of sparse dimensions in `self` ",
+              "should match that of the `mask`. ",
+              "Got `self.sparse_dim() == ", t.sparse_dim(), "` != ",
+              "`mask.sparse_dim() == ", mask.sparse_dim(), "`.");
+
+  const auto wrapped_tensor = [](const Tensor& t,
+                                 const OptTensor& indices = c10::nullopt,
+                                 const OptTensor& values = c10::nullopt) -> Tensor {
+    auto res = at::empty({0}, t.options());
+    auto* res_sparse_impl = get_sparse_impl(res);
+    res_sparse_impl->raw_resize_(t.sparse_dim(), t.dense_dim(), t.sizes());
+    const auto res_indices = indices.has_value() ? *indices : t._indices();
+    const auto res_values = values.has_value() ? *values : t._values();
+    res_sparse_impl->set_indices_and_values_unsafe(res_indices, res_values);
+    res_sparse_impl->set_nnz_and_narrow(t._nnz());
+    res._coalesced_(false);
+    return res;
+  };
+
+  Tensor lhs;
+  OptTensor lhs_hash_opt;
+
+  std::tie(lhs, lhs_hash_opt) = [&]() -> auto {
+    if (t.is_coalesced()) {
+      return std::make_tuple(t, static_cast<OptTensor>(c10::nullopt));
+    } else {
+      const auto indices_hash = at::sparse::flatten_indices(t._indices(), t.sizes());
+      const auto argsort_indices_hash = std::get<1>(indices_hash.sort(0));
+      // Probably worth having a dedicated kernel for.
+      const auto res_indices = t._indices().index_select(1, argsort_indices_hash);
+      const auto res_values = t._values().index_select(0, argsort_indices_hash);
+      const auto indices_hash_sorted = indices_hash.index_select(0, argsort_indices_hash);
+      // NOTE: res is not necessariy coalesced, but it is sorted.
+      // We mark it as "coalesced" to skip sorting in the intersection kernel.
+      auto res = wrapped_tensor(t, res_indices, res_values)._coalesced_(true);
+      return std::make_tuple(res, static_cast<OptTensor>(indices_hash_sorted));
+    }
+  }();
+
+  const auto rhs = mask.is_coalesced() ? wrapped_tensor(mask) : mask;
+
+  return std::make_tuple(lhs, rhs, lhs_hash_opt);
+}
 
 SparseTensor sparse_mask(const Tensor& t, const SparseTensor& mask) {
   TORCH_CHECK(
```

Comment:
This check is needed because not all inputs are sparse, and `sparse_mask_like_prepare_sparse_inputs` works only with sparse inputs. I could remove the shape check in `sparse_mask_like_prepare_sparse_inputs` instead.